File: scripts/check-bench-regression.py
#!/usr/bin/env python3
# SPDX-License-Identifier: PMPL-1.0-or-later
# hypatia:ignore cicd_rules/banned_language_file
# Intentional exception to the org's no-Python policy: this script
# parses criterion's bencher-format output and compares against the
# committed baseline. Pairs with update-bench-baselines.py.
# Tracked in the .hypatia-exemptions.md table.
"""
check-bench-regression.py — compare a criterion bencher run against
.machine_readable/benchmarks/baselines.json and fail if any benchmark
has regressed by more than the configured threshold.

Usage:
    python3 scripts/check-bench-regression.py \\
        /tmp/bench.txt \\
        .machine_readable/benchmarks/baselines.json

Exit status:
    0 — no regressions exceed the threshold (or baseline is empty).
    1 — at least one benchmark exceeds the threshold.
    2 — usage / file error.

Writes a Markdown-formatted summary to stdout suitable for
`$GITHUB_STEP_SUMMARY`. Emits `::error::` annotations for regressions
so GitHub annotates the offending job.
"""

from __future__ import annotations

import json
import re
import sys
from pathlib import Path


BENCH_LINE = re.compile(
    r"^test\s+(?P<name>\S+)\s+\.\.\.\s+bench:\s+(?P<ns>[\d,]+)\s+ns/iter"
)


def parse_bencher_output(text: str) -> dict[str, int]:
    out: dict[str, int] = {}
    for line in text.splitlines():
        m = BENCH_LINE.match(line.strip())
        if m:
            out[m.group("name")] = int(m.group("ns").replace(",", ""))
    return out


def fmt_ns(ns: int) -> str:
    if ns >= 1_000_000:
        return f"{ns / 1_000_000:.2f} ms"
    if ns >= 1_000:
        return f"{ns / 1_000:.2f} µs"
    return f"{ns} ns"


def main(argv: list[str]) -> int:
    if len(argv) != 3:
        print(
            "usage: check-bench-regression.py <bencher-output> <baselines.json>",
            file=sys.stderr,
        )
        return 2

    current_path = Path(argv[1])
    baselines_path = Path(argv[2])

    if not current_path.exists():
        print(f"error: {current_path} missing", file=sys.stderr)
        return 2

    current = parse_bencher_output(current_path.read_text())

    if not current:
        print(
            "::warning::no bench lines parsed from current run — "
            "did criterion use --output-format bencher?"
        )
        return 0

    baseline_doc = {}
    if baselines_path.exists():
        try:
            baseline_doc = json.loads(baselines_path.read_text())
        except json.JSONDecodeError:
            print(
                f"::warning::{baselines_path} is not valid JSON; "
                "treating as empty baseline"
            )

    baselines = baseline_doc.get("baselines", {}) or {}
    threshold_pct = float(baseline_doc.get("_regression_threshold_pct", 50))

    if not baselines:
        print("## Benchmark run (advisory mode — no baselines yet)")
        print()
        print("| Benchmark | Current |")
        print("|-----------|---------|")
        for name, ns in sorted(current.items()):
            print(f"| `{name}` | {fmt_ns(ns)} |")
        print()
        print(
            "_No entries in `baselines.json` yet — see "
            "`.machine_readable/benchmarks/README.md` for how to seed them._"
        )
        return 0

    # Compare
    regressions: list[tuple[str, int, int, float]] = []
    report_rows: list[tuple[str, str, str, str, str]] = []

    for name, ns_now in sorted(current.items()):
        ns_base = baselines.get(name)
        if ns_base is None:
            report_rows.append(
                (name, fmt_ns(ns_now), "—", "new", "✨")
            )
            continue

        pct = (ns_now - ns_base) / ns_base * 100 if ns_base else 0.0
        verdict = "✅"
        if pct > threshold_pct:
            verdict = "❌"
            regressions.append((name, ns_base, ns_now, pct))
        elif pct > threshold_pct / 2:
            verdict = "⚠️"
        elif pct < -10:
            verdict = "🚀"

        report_rows.append(
            (name, fmt_ns(ns_now), fmt_ns(ns_base), f"{pct:+.1f}%", verdict)
        )

    print("## Benchmark comparison")
    print()
    print(f"Threshold: regression > **{threshold_pct:.0f}%** fails CI.")
    print()
    print("| Benchmark | Current | Baseline | Δ | |")
    print("|-----------|---------|----------|---|---|")
    for row in report_rows:
        print(f"| `{row[0]}` | {row[1]} | {row[2]} | {row[3]} | {row[4]} |")
    print()

    if regressions:
        print("### Regressions exceeding threshold")
        print()
        for name, ns_base, ns_now, pct in regressions:
            msg = (
                f"{name}: {fmt_ns(ns_base)} → {fmt_ns(ns_now)} "
                f"({pct:+.1f}%, threshold {threshold_pct:.0f}%)"
            )
            print(f"- {msg}")
            # GitHub annotation on stderr so the summary on stdout stays clean.
            print(f"::error::benchmark regression: {msg}", file=sys.stderr)
        return 1

    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: scripts/bench-tools/src/bin/check-bench-regression.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
//
// check-bench-regression — compare a criterion bencher run against
// .machine_readable/benchmarks/baselines.json and fail if any benchmark
// regressed by more than the configured threshold. A faithful Rust port of
// the former scripts/check-bench-regression.py (org policy bans Python
// outside SaltStack). Pairs with update-bench-baselines.
//
// Usage:
//     check-bench-regression <bencher-output> <baselines.json>
//
// Exit status:  0 = no regressions over threshold (or no baselines yet),
//               1 = at least one regression,  2 = usage / file error.
//
// Markdown summary -> stdout (for $GITHUB_STEP_SUMMARY); `::error::`
// annotations -> stderr.

use bench_tools::{fmt_ns, parse_bencher_output, parse_json, Json};
use std::process::exit;

fn main() {
    let argv: Vec<String> = std::env::args().collect();
    if argv.len() != 3 {
        eprintln!("usage: check-bench-regression <bencher-output> <baselines.json>");
        exit(2);
    }
    let current_path = &argv[1];
    let baselines_path = &argv[2];

    let current_text = match std::fs::read_to_string(current_path) {
        Ok(t) => t,
        Err(_) => {
            eprintln!("error: {current_path} missing");
            exit(2);
        }
    };

    let mut current = parse_bencher_output(&current_text);
    current.sort_by(|a, b| a.0.cmp(&b.0)); // Python iterates `sorted(current.items())`

    if current.is_empty() {
        println!(
            "::warning::no bench lines parsed from current run \u{2014} \
             did criterion use --output-format bencher?"
        );
        exit(0);
    }

    let baseline_doc: Json = match std::fs::read_to_string(baselines_path) {
        Ok(t) => match parse_json(&t) {
            Ok(v) => v,
            Err(_) => {
                println!(
                    "::warning::{baselines_path} is not valid JSON; \
                     treating as empty baseline"
                );
                Json::Obj(vec![])
            }
        },
        Err(_) => Json::Obj(vec![]),
    };

    let baselines: Vec<(String, f64)> = match baseline_doc.get("baselines") {
        Some(Json::Obj(p)) => p
            .iter()
            .filter_map(|(k, v)| v.as_f64().map(|n| (k.clone(), n)))
            .collect(),
        _ => vec![],
    };
    let lookup = |name: &str| baselines.iter().find(|(k, _)| k == name).map(|(_, v)| *v);

    let threshold_pct = baseline_doc
        .get("_regression_threshold_pct")
        .and_then(|v| v.as_f64())
        .unwrap_or(50.0);

    if baselines.is_empty() {
        println!("## Benchmark run (advisory mode \u{2014} no baselines yet)");
        println!();
        println!("| Benchmark | Current |");
        println!("|-----------|---------|");
        for (name, ns) in &current {
            println!("| `{name}` | {} |", fmt_ns(*ns));
        }
        println!();
        println!(
            "_No entries in `baselines.json` yet \u{2014} see \
             `.machine_readable/benchmarks/README.md` for how to seed them._"
        );
        exit(0);
    }

    let mut regressions: Vec<(String, i64, i64, f64)> = vec![];
    let mut rows: Vec<(String, String, String, String, String)> = vec![];

    for (name, ns_now) in &current {
        let ns_now = *ns_now;
        match lookup(name) {
            None => rows.push((
                name.clone(),
                fmt_ns(ns_now),
                "\u{2014}".into(),
                "new".into(),
                "\u{2728}".into(),
            )),
            Some(ns_base) => {
                let pct = if ns_base != 0.0 {
                    (ns_now as f64 - ns_base) / ns_base * 100.0
                } else {
                    0.0
                };
                let mut verdict = "\u{2705}";
                if pct > threshold_pct {
                    verdict = "\u{274c}";
                    regressions.push((name.clone(), ns_base as i64, ns_now, pct));
                } else if pct > threshold_pct / 2.0 {
                    verdict = "\u{26a0}\u{fe0f}";
                } else if pct < -10.0 {
                    verdict = "\u{1f680}";
                }
                rows.push((
                    name.clone(),
                    fmt_ns(ns_now),
                    fmt_ns(ns_base as i64),
                    format!("{pct:+.1}%"),
                    verdict.into(),
                ));
            }
        }
    }

    println!("## Benchmark comparison");
    println!();
    println!("Threshold: regression > **{threshold_pct:.0}%** fails CI.");
    println!();
    println!("| Benchmark | Current | Baseline | \u{0394} | |");
    println!("|-----------|---------|----------|---|---|");
    for (a, b, c, d, e) in &rows {
        println!("| `{a}` | {b} | {c} | {d} | {e} |");
    }
    println!();

    if !regressions.is_empty() {
        println!("### Regressions exceeding threshold");
        println!();
        for (name, ns_base, ns_now, pct) in &regressions {
            let msg = format!(
                "{name}: {} \u{2192} {} ({pct:+.1}%, threshold {threshold_pct:.0}%)",
                fmt_ns(*ns_base),
                fmt_ns(*ns_now),
            );
            println!("- {msg}");
            eprintln!("::error::benchmark regression: {msg}");
        }
        exit(1);
    }
}
